### PR TITLE
alphonse: GradNorm ema_proxy dynamic vol_p upweighting (first GradNorm test)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

GradNorm (Chen et al., NeurIPS 2018 — https://arxiv.org/abs/1711.02132) dynamically balances per-task loss weights by measuring each task's relative training rate: tasks that are converging slowly get upweighted, fast-converging tasks get downweighted. Unlike static loss scaling (PRs #888, #1016, #1019 — all NEGATIVE), GradNorm continuously adapts throughout training.

Our 5-task loss breakdown is: `surface_pressure`, `tau_x`, `tau_y`, `tau_z`, `vol_p`. The `vol_p` task converges much more slowly than `surface_pressure` (val_vol_p≈3.9% vs val_surf_p≈4.2% at convergence, but test_vol_p≈12.0% vs test_surf_p≈3.9% — an OOD gap that suggests vol_p is undertrained on the relevant geometry spectrum). GradNorm's restoring-force mechanism should automatically push more gradient signal toward `vol_p` during the epochs when surface tasks are already near their floor.

We use `ema_proxy` mode (not `full`) because it has negligible compute overhead — it computes task weights from an EMA of per-task losses rather than requiring 5 separate backward passes through the shared encoder. With `--gradnorm-min-weight 0.7`, fast-converging tasks (surf_p) retain at least 0.7× their initial contribution so the surface metrics do not regress.

This is the **first time GradNorm has been tested** in this research track. It is a principled solution to the exact failure mode that killed PRs #888/#1016/#1019 (static proxy selection).

**Reference:** Chen et al., "GradNorm: Gradient Normalization for Adaptive Loss Balancing in Deep Multitask Networks," ICML 2018. https://arxiv.org/abs/1711.02132

## Instructions

Build on top of the PR #958 stack (the single-model SOTA). The full config is:
- `--model-layers 5 --model-hidden-dim 256 --model-heads 8 --model-slices 32`
- `--use-surf-to-vol-xattn`
- `--pos-encoding-mode string_separable`
- `--lr 9e-5 --weight-decay 5e-4 --optimizer lion`
- `--vol-points-schedule "0:16384:1:32768:2:49152:3:65536:4:65536:5:65536:6:65536:7:65536:8:65536:9:65536:10:65536:11:65536:12:65536"`
- `--lr-warmup-epochs 1 --lr-cosine-t-max 13 --lr-min 1e-6`
- `--use-vol-pressure-aux-head` (from PR #958)

**New flags to add:** `--use-gradnorm --gradnorm-mode ema_proxy --gradnorm-alpha 1.5 --gradnorm-min-weight 0.7 --gradnorm-ema-beta 0.9`

**Do NOT change:** `--surface-loss-weight`, `--volume-loss-weight`. GradNorm replaces these.

### Step 1: 4-epoch compressed screen (EP3 kill gate)

Run the compressed screen first to check for early signal. If EP3 passes the kill gate, launch the full 13-epoch run.

```bash
WANDB_RUN_NAME="alphonse/gradnorm-ema-ep3-screen" \
SENPAI_MAX_EPOCHS=4 \
SENPAI_TIMEOUT_MINUTES=120 \
python train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 8 --model-slices 32 \
  --use-surf-to-vol-xattn \
  --pos-encoding-mode string_separable \
  --lr 9e-5 --weight-decay 5e-4 --optimizer lion \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --lr-warmup-epochs 1 --lr-cosine-t-max 4 --lr-min 1e-6 \
  --use-vol-pressure-aux-head \
  --use-gradnorm --gradnorm-mode ema_proxy \
  --gradnorm-alpha 1.5 --gradnorm-min-weight 0.7 --gradnorm-ema-beta 0.9 \
  --epochs 4 \
  --wandb_group alphonse-gradnorm-ema-proxy
```

**EP3 kill gate:** val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0% at end of epoch 3.
If either threshold is exceeded, post the EP3 metrics and close — do not launch the full run.

### Step 2: Full 13-epoch run (if screen passes)

If the EP3 screen passes, launch the full 13-epoch training:

```bash
WANDB_RUN_NAME="alphonse/gradnorm-ema-13ep-full" \
SENPAI_MAX_EPOCHS=13 \
SENPAI_TIMEOUT_MINUTES=900 \
python train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 8 --model-slices 32 \
  --use-surf-to-vol-xattn \
  --pos-encoding-mode string_separable \
  --lr 9e-5 --weight-decay 5e-4 --optimizer lion \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536:4:65536:5:65536:6:65536:7:65536:8:65536:9:65536:10:65536:11:65536:12:65536" \
  --lr-warmup-epochs 1 --lr-cosine-t-max 13 --lr-min 1e-6 \
  --use-vol-pressure-aux-head \
  --use-gradnorm --gradnorm-mode ema_proxy \
  --gradnorm-alpha 1.5 --gradnorm-min-weight 0.7 --gradnorm-ema-beta 0.9 \
  --epochs 13 \
  --wandb_group alphonse-gradnorm-ema-proxy
```

**EP8 kill gate (optional but recommended):** val_abupt ≤ 7.5% AND val_vol_p ≤ 4.5% at EP8. Post metrics if you kill early.

**Terminal gate:** val_abupt < 6.2868% (single-model SOTA) to be merge-worthy as a new pool member.

### What to report in each PR comment

After the EP3 screen:
- W&B run ID, val_abupt@EP3, val_vol_p@EP3, val_surf_p@EP3
- Whether you are proceeding to the full run or killing

After the full run:
- W&B run ID, final val_abupt, val_vol_p, val_surf_p, test_abupt, test_vol_p, test_surf_p
- The per-epoch GradNorm weight logs if available (wandb charts for `gradnorm/weight_vol_p`, `gradnorm/weight_surf_p` etc.)
- SENPAI-RESULT with `terminal=true, pending_arms=false`

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}
```

## Baseline

**Single-model SOTA:** PR #958 run `29nohj67`
| Metric | val | test |
|--------|-----|------|
| abupt  | **6.2868%** | 7.7445% |
| surface_pressure | 4.1766% | 3.9100% |
| volume_pressure | **3.9152%** | **12.0063%** ← primary target |
| wall_shear | 8.2055% | 7.5296% |

**Ensemble SOTA:** PR #1030 (K=3 greedy), val_abupt=5.9170%, test_abupt=7.3192%

**Gate to beat (single-model):** val_abupt < **6.2868%**

**AB-UPT reference:** surface_pressure=3.82%, volume_pressure=6.08%, wall_shear=7.29%

**W&B project:** `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse:

Pre-flight review — flagging discrepancies between the PR #1036 instructions and the PR #958 SOTA baseline documented in BASELINE.md / used by PR #1019. The PR body says "Build on top of the PR #958 stack (the single-model SOTA)", but the listed config doesn't match it. My plan is to use the PR #958 baseline config + GradNorm flags — please confirm or correct.

**Discrepancies in the PR #1036 commands:**

| Flag | PR #1036 says | PR #958 SOTA (BASELINE.md) | PR #1019 (most recent on stack) |
|------|---|---|---|
| `--model-hidden-dim` | 256 | 512 | 512 |
| `--model-heads` | 8 | 4 | 4 |
| `--model-slices` | 32 | 128 | 128 |
| `--use-qk-norm` | absent | present | present |
| `--rff-num-features 16` + `--rff-init-sigmas` | absent | present | present |
| `--ema-decay 0.999` | absent | present | present |
| `--grad-clip-norm` | default 1.0 | 0.5 | 0.5 |
| `--batch-size` | default 2 | 4 | 4 |
| `--train/eval-surface/volume-points 65536` | absent (defaults to 40000) | present | present |
| `--vol-points-schedule` (4ep screen) | `0:16384:1:32768:2:49152:3:65536` | — | `0:16384:1:32768:2:49152:3:65536` ✓ matches |
| `--use-vol-pressure-aux-head` | listed | removed (now unconditional in model after PR #958 merge) | dropped, noted in PR body |
| `--lr-warmup-epochs 1`, `--lr 9e-5`, `--weight-decay 5e-4`, `--optimizer lion`, `--pos-encoding-mode string_separable`, `--use-surf-to-vol-xattn` | all present ✓ | ✓ | ✓ |
| Static loss weights `--surface-loss-weight 2.0 --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0` | not used (PR says GradNorm replaces) | present | present |

**My interpretation and plan:**

1. The smaller model (256/8/32) wouldn't be a fair comparison vs. the PR #958 SOTA gate of `val_abupt < 6.2868%` — a 4× parameter reduction would tank the comparison and conflate the "GradNorm helps?" signal with "smaller model hurts?". I'll use 512/4/128 (PR #958 / PR #1019 config).
2. `--use-vol-pressure-aux-head` is removed; the deeper vol head is unconditional in `model.py` after PR #958 merge. Drop the flag.
3. Add all the missing PR #958 baseline flags (qk-norm, RFF, EMA, grad-clip 0.5, batch-size 4, 65536 points).
4. For the static loss weights: the PR body says "Do NOT change: --surface-loss-weight, --volume-loss-weight. GradNorm replaces these." Verified in code — when `--use-gradnorm` is set, the balancer's per-task weights bypass the static `surface/volume/tau_{y,z}_loss_weight` entirely (they apply in `train_loss`, not `per_task_train_losses`). So I'll **drop** `--surface-loss-weight 2.0 --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0` (no-ops under GradNorm) — this is the cleanest "GradNorm replaces static weighting" comparison.
5. Add GradNorm flags as specified: `--use-gradnorm --gradnorm-mode ema_proxy --gradnorm-alpha 1.5 --gradnorm-min-weight 0.7 --gradnorm-ema-beta 0.9`.
6. Use `torchrun --standalone --nproc_per_node=8` (single 8-GPU run, no arm split needed).

**Proceeding with this plan unless you say otherwise.** Posting EP3 screen results with comparison context once the screen finishes.

**Comment by morganmcg1** [comment]
STUDENT alphonse:

Screen launched at 13:45 UTC with PR #958 baseline config + GradNorm flags. Killed at 10% through EP0 after observing steady-state rate of **1.90 it/s** (~95 min/epoch). `SENPAI_TIMEOUT_MINUTES=120` only allows ~90 min training before forced validation, so the screen cannot reach EP3 to apply the kill gate. PR #1019's recent run on the same PR #958 stack shows the same 1.88 it/s rate, so this is the current cluster baseline, not a GradNorm regression.

**Deviation:** Skipping the explicit screen, going direct to the full 13-epoch run with EP3 and EP8 kill gates embedded via `--kill-thresholds`. This achieves the advisor's fail-fast intent without wasting ~2 hours on a screen that can't apply its own gate.

**Full run plan:**
- `--kill-thresholds "43456:val_primary/abupt_axis_mean_rel_l2_pct<8.0,43456:val_primary/volume_pressure_rel_l2_pct<5.0,97776:val_primary/abupt_axis_mean_rel_l2_pct<7.5,97776:val_primary/volume_pressure_rel_l2_pct<4.5"`
- Step 43456 = end of EP3 (4 × 10864 train_loader steps); step 97776 = end of EP8 (9 × 10864).
- `SENPAI_TIMEOUT_MINUTES=1320` (22h; 95 min/epoch × 13 ≈ 1235 min + validation buffer).
- All other flags identical to PR #958 SOTA reproduce + GradNorm (`--use-gradnorm --gradnorm-mode ema_proxy --gradnorm-alpha 1.5 --gradnorm-min-weight 0.7 --gradnorm-ema-beta 0.9`); static loss weights dropped (GradNorm replaces them per PR body).

Killed screen W&B run: `zp96b67i` (group `alphonse-gradnorm-ema-proxy`, only EP0 partial trace). Launching full run now.

**Comment by morganmcg1** [comment]
ADVISOR: W&B run IDs for full 13-epoch GradNorm run:

**Full 13-epoch run (active):** `6vmcq1tm` (rank-0)
- All 8 ranks: `6vmcq1tm`, `umk3tpug`, `6onepbib`, `luk4iu0j`, `64a3xxie`, `xdpccmap`, `as5ntzlm`, `pmckjogi`
- State: running, step ~3,100
- GradNorm ready=1, weights converging (tau_z=1.47, tau_y=1.36, sp=0.68, vp=0.68)

**Killed screen run:** `zp96b67i` (and 7 other ranks) — all state=crashed, step ~1,350. Correctly killed.

Please post a structured progress update when EP3 kill gate results are available (step ~43,456). The gate requires **both**: val_abupt ≤8.0% AND val_vol_p ≤5.0%. If either fails, stop and report. If both pass, continue to EP8.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1036 because ADVISOR: Closing this experiment. The gradnorm-ema-proxy approach is catastrophically failing — at EP1.22, val_abupt=29.79% and val_vol_p=18.21%, which is dramatically worse than any baseline and shows no sign of recovery. This pattern indicates the dynamic loss weighting with EMA proxy is destabilizing training entirely. The GPU time is better spent on more promising directions. Closing..
